File: src/nl/rubensten/texifyidea/index/stub/LatexCommandsStubElementType.java
package nl.rubensten.texifyidea.index.stub;

import com.intellij.psi.stubs.*;
import nl.rubensten.texifyidea.LatexLanguage;
import nl.rubensten.texifyidea.index.LatexCommandsIndex;
import nl.rubensten.texifyidea.index.LatexDefinitionIndex;
import nl.rubensten.texifyidea.index.LatexIncludesIndex;
import nl.rubensten.texifyidea.psi.LatexCommands;
import nl.rubensten.texifyidea.psi.impl.LatexCommandsImpl;
import nl.rubensten.texifyidea.util.Magic;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.List;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

/**
 * @author Ruben Schellekens
 */
public class LatexCommandsStubElementType extends IStubElementType<LatexCommandsStub, LatexCommands> {

    private static final Pattern SEPERATOR = Pattern.compile("\u1923\u9123\u2d20 hello\u0012");

    public LatexCommandsStubElementType(@NotNull String debugName) {
        super("latex-commands", LatexLanguage.INSTANCE);
    }

    @Override
    public LatexCommands createPsi(@NotNull LatexCommandsStub latexCommandsStub) {
        return new LatexCommandsImpl(latexCommandsStub, this) {{
            setName(latexCommandsStub.getName());
        }};
    }

    @NotNull
    @Override
    public LatexCommandsStub createStub(@NotNull LatexCommands latexCommands, StubElement parent) {
        String commandToken = latexCommands.getCommandToken().getText();
        latexCommands.setName(commandToken);

        List<String> requiredParameters = latexCommands.getRequiredParameters();
        List<String> optionalParameters = latexCommands.getOptionalParameters();

        return new LatexCommandsStubImpl(
                parent, this,
                commandToken,
                requiredParameters,
                optionalParameters
        );
    }

    @NotNull
    @Override
    public String getExternalId() {
        return "texify.latex.commands";
    }

    @Override
    public void serialize(@NotNull LatexCommandsStub latexCommandsStub, @NotNull StubOutputStream stubOutputStream) throws IOException {
        stubOutputStream.writeName(latexCommandsStub.getName());
        stubOutputStream.writeName(serialiseRequired(latexCommandsStub));
        stubOutputStream.writeName(serialiseOptional(latexCommandsStub));
    }

    @NotNull
    @Override
    public LatexCommandsStub deserialize(@NotNull StubInputStream stubInputStream, StubElement parent) throws IOException {
        final String name = stubInputStream.readName().toString();
        final List<String> required = deserialiseList(stubInputStream.readName().toString());
        final List<String> optional = deserialiseList(stubInputStream.readName().toString());

        return new LatexCommandsStubImpl(
                parent, this,
                name,
                required,
                optional
        );
    }

    @Override
    public void indexStub(@NotNull LatexCommandsStub latexCommandsStub, @NotNull IndexSink indexSink) {
        indexSink.occurrence(LatexCommandsIndex.Companion.key(), latexCommandsStub.getCommandToken());

        String token = latexCommandsStub.getCommandToken();
        if (Magic.Command.includes.contains(token)) {
            indexSink.occurrence(LatexIncludesIndex.Companion.key(), token);
        }

        if (Magic.Command.definitions.contains(token) || Magic.Command.redefinitions.contains(token)) {
            indexSink.occurrence(LatexDefinitionIndex.Companion.key(), token);
        }
    }

    @NotNull
    private List<String> deserialiseList(@NotNull String string) {
        return SEPERATOR.splitAsStream(string).collect(Collectors.toList());
    }

    @NotNull
    private String serialiseRequired(@NotNull LatexCommandsStub stub) {
        return String.join(SEPERATOR.pattern(), stub.getRequiredParams());
    }

    @NotNull
    private String serialiseOptional(@NotNull LatexCommandsStub stub) {
        return String.join(SEPERATOR.pattern(), stub.getOptionalParams());
    }
}

File: src/nl/hannahsten/texifyidea/index/stub/LatexCommandsStubElementType.kt
package nl.hannahsten.texifyidea.index.stub

import com.intellij.psi.stubs.*
import nl.hannahsten.texifyidea.grammar.LatexLanguage
import nl.hannahsten.texifyidea.index.LatexStubIndexKeys
import nl.hannahsten.texifyidea.index.NewDefinitionIndex
import nl.hannahsten.texifyidea.index.NewLabelsIndex
import nl.hannahsten.texifyidea.index.NewSpecialCommandsIndex
import nl.hannahsten.texifyidea.psi.LatexCommands
import nl.hannahsten.texifyidea.psi.LatexParameter
import nl.hannahsten.texifyidea.psi.contentText
import nl.hannahsten.texifyidea.psi.impl.LatexCommandsImpl
import nl.hannahsten.texifyidea.util.parser.forEachDirectChildTyped
import nl.hannahsten.texifyidea.util.parser.toStringMap
import java.io.IOException
import java.util.regex.Pattern
import java.util.stream.Collectors

/**
 * @author Hannah Schellekens
 */
class LatexCommandsStubElementType(debugName: String) :
    IStubElementType<LatexCommandsStub, LatexCommands>(debugName, LatexLanguage) {

    override fun createPsi(latexCommandsStub: LatexCommandsStub): LatexCommands = LatexCommandsImpl(latexCommandsStub, this)

    override fun createStub(latexCommands: LatexCommands, parent: StubElement<*>): LatexCommandsStub {
        val commandToken = latexCommands.commandToken.text
        val parameterContents = mutableListOf<LatexParameterStub>()
        latexCommands.forEachDirectChildTyped<LatexParameter> {
            val text = it.contentText()
            val type = if (it.requiredParam != null) LatexParameterStub.REQUIRED
            else if (it.optionalParam != null) LatexParameterStub.OPTIONAL else return@forEachDirectChildTyped
            parameterContents.add(LatexParameterStub(type, text))
        }
        val optionalParametersMap: Map<String, String> =
            latexCommands.getOptionalParameterMap().toStringMap()
        return LatexCommandsStubImpl(
            parent, this,
            commandToken, parameterContents, optionalParametersMap
        )
    }

    // Should equal externalIdPrefix from registration in index.xml plus field name in LatexStubElementTypes
    override fun getExternalId() = "texify.latex." + super.toString()

    private fun writeParameters(
        parameters: List<LatexParameterStub>,
        output: StubOutputStream
    ) {
        output.writeVarInt(parameters.size)
        parameters.forEach { parameter ->
            output.writeByte(parameter.type)
            output.writeUTFFast(parameter.content)
        }
    }

    private fun readParameters(input: StubInputStream): List<LatexParameterStub> {
        val size = input.readVarInt()
        return (0 until size).map {
            val type = input.readByte().toInt()
            val content = input.readUTFFast()
            LatexParameterStub(type, content)
        }
    }

    @Throws(IOException::class)
    override fun serialize(
        latexCommandsStub: LatexCommandsStub,
        stubOutputStream: StubOutputStream
    ) {
        stubOutputStream.writeName(latexCommandsStub.name)
        writeParameters(latexCommandsStub.parameters, stubOutputStream)
        stubOutputStream.writeName(serialiseOptionalMap(latexCommandsStub))
    }

    @Throws(IOException::class)
    override fun deserialize(stubInputStream: StubInputStream, parent: StubElement<*>): LatexCommandsStub {
        val name = stubInputStream.readName().toString()
        val parameters = readParameters(stubInputStream)
        val optional = deserializeMap(stubInputStream.readName().toString())
        return LatexCommandsStubImpl(
            parent, this, name, parameters, optional
        )
    }

    private fun deserializeMap(fromString: String): Map<String, String> {
        if(fromString.isEmpty()) return emptyMap()
        val keyValuePairs = deserialiseList(fromString)
        return keyValuePairs.filter { it.isNotEmpty() }.associate {
            val parts = it.split(KEY_VALUE_SEPARATOR)
            parts[0] to parts[1]
        }
    }

    override fun indexStub(stub: LatexCommandsStub, sink: IndexSink) {
        val token = stub.commandToken
        sink.occurrence(LatexStubIndexKeys.COMMANDS, token)

        NewSpecialCommandsIndex.sinkIndex(sink, token) // all commands classified
        NewDefinitionIndex.sinkIndex(stub, sink) // record definitions
        NewLabelsIndex.sinkIndexCommand(stub, sink) // labels
    }

    private fun deserialiseList(string: String): List<String> = LIST_ELEMENT_SEPARATOR.splitAsStream(string)
        .collect(Collectors.toList())

    private fun serialiseOptionalMap(stub: LatexCommandsStub): String {
        val keyValuePairs = stub.optionalParamsMap.map { "${it.key}=${it.value}" }
        return java.lang.String.join(LIST_ELEMENT_SEPARATOR.pattern(), keyValuePairs)
    }

    companion object {

        private val LIST_ELEMENT_SEPARATOR =
            Pattern.compile("\u1923\u9123\u2d20 hello\u0012")
        private val KEY_VALUE_SEPARATOR = "=".toRegex()
    }
}